### PR TITLE
Calculator: Move towards a Calculator without any numerical errors (part 1)

### PR DESCRIPTION
**Pull request by LucasChollet**
I'm splitting #11623 in multiple PR to ease the review process.

This PR includes all following changes:

- [x] Support any number
- [x] Support flawless +, -, *, /, inv
- [x] Display rounded numbers
- [ ] Enhance user interaction with rounding

### CODE REVIEW

File: Userland/Applications/Calculator/Keypad.cpp [hunk 1]
```diff
@@ -68,70 +60,80 @@ void Keypad::type_backspace()
...
-KeypadValue Keypad::value() const
+Crypto::BigFraction Keypad::value() const
 {
-    KeypadValue frac_part = { (i64)m_frac_value.value(), m_frac_length };
-    KeypadValue int_part = { (i64)m_int_value.value() };
-    KeypadValue res = int_part + frac_part;
-    if (m_negative)
-        res = -res;
-    return res;
+    using Crypto::NumberTheory::Power;
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/Calculator/Keypad.cpp:86]: Unnecessary

File: Userland/Applications/Calculator/Keypad.h [hunk 2]
```diff
@@ -24,23 +25,27 @@ class Keypad final {
...
     String to_string() const;
 
 private:
     // Internal representation of the current decimal value.
-    bool m_negative { false };
-    Checked<u64> m_int_value { 0 };
-    Checked<u64> m_frac_value { 0 };
-    u8 m_frac_length { 0 };
+    // Those variables are only used when the user is entering a value.
+    // Otherwise, the KeypadValue m_internal_value is used.
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/Calculator/Keypad.h:37]: It's a BigFraction now

File: Userland/Applications/Calculator/main.cpp [hunk 3]
```diff
@@ -51,20 +52,22 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
         if (clipboard.mime_type == "text/plain") {
             if (!clipboard.data.is_empty()) {
-                widget->set_entry(KeypadValue(StringView(clipboard.data)));
+                widget->set_entry(Crypto::BigFraction(StringView(clipboard.data)));
             }
         }
     }));
 
     auto& constants_menu = window->add_menu("&Constants");
+    auto const power = Crypto::NumberTheory::Power("10"_bigint, "10"_bigint);
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/Calculator/main.cpp:61]: Doesn't that change the constant's values further down?

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]

File: Userland/Applications/Calculator/main.cpp (hunk 3 above)
Comment (LucasChollet) [reply Userland/Applications/Calculator/main.cpp:61]: In what way?
I can't re-test it now, but I remember testing it.

File: Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+BigFraction BigFraction::operator-(BigFraction const& rhs) const
+{
+    return *this + (-rhs);
+}
+
+BigFraction BigFraction::operator*(BigFraction const& rhs) const
+{
+    auto tmp = *this;
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp:65]: `tmp` again

File: Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+BigFraction BigFraction::invert() const
+{
+    return BigFraction { 1 } / *this;
+}
+
+BigFraction BigFraction::operator/(BigFraction const& rhs) const
+{
+    VERIFY(rhs.m_numerator != "0"_bigint);
+
+    auto tmp = *this;
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp:88]: `tmp` again

File: Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+    auto const remove_tailing_zeros = [](StringView value) -> StringView {
+        auto n = value.length();
+        VERIFY(n > 0);
+        while (value.characters_without_null_termination()[n - 1] == '0')
+            --n;
+        return { value.characters_without_null_termination(), n };
+    };
+
+    auto const raw_frac_value = full_value.substring(split, full_value.length() - split);
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp:220]: From the coding style:
> Use full words, except in the rare case where an abbreviation would be more canonical and easier to understand.

So, "raw_frac_value" -> "raw_fractional_value"

File: Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+
+    auto const integer_value = split == 0 ? "0"sv : full_value.substring_view(0, split);
+    auto const frac_value = rounding_threshold == 0 ? "0"sv : remove_tailing_zeros(raw_frac_value);
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp:223]: "frac_value" -> "fractional_value"

File: Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    auto const rounded_fraction = rounded(rounding_threshold);
+
+    // We take the unsigned value as we already manage the '-'
+    auto const full_value = rounded_fraction.m_numerator.unsigned_value().to_base(10);
+    int split = full_value.length() - (number_of_digits(rounded_fraction.m_denominator) - 1);
+
+    if (split < 0)
+        split = 0;
+
+    auto const remove_tailing_zeros = [](StringView value) -> StringView {
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp:212]: I think you mean "trailing" zeros not "tailing"?

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Finally got over my fear of _**mathematics**_ (:scream:) and had a look through this, sorry it's sat for so long! Seems good, I just have some nit-picks about naming things and then it should be good to merge. :^)

File: Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp (hunk 8 above)
Comment (LucasChollet) [reply Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp:212]: That's a cute mistake ahah

**Comment by LucasChollet** [comment]
Hope the experience wasn't that terrible! Thank you very much for reviewing this. It was my first "big" PR for SerenityOS, I will be very happy to see it merged.

I also had to remove all `BigInteger::create_from()` occurrences and replace them with direct call to the constructor (due to #15029).

**Review by AtkinsSJ — APPROVED** [review]
LGTM! Just want to let CI finish.

**Comment by LucasChollet** [comment]
Sorry for the mess, I'm trying to solve the macOS CI issue.

**Comment by AtkinsSJ** [comment]
The issue seems to be that it doesn't think `size_t` and `u64` are the same type. (They should be, unless the Mac build is 32 bit?) It sees an overload for `u64`, and one for `double`, but `size_t` is neither of those so it can't decide.

I think you can fix it by changing `BigFraction.cpp:39` from

```c++
auto fraction_length = UnsignedBigInteger(fraction_part_view.length());
```

to

```c++
auto fraction_length = UnsignedBigInteger(static_cast<u64>(fraction_part_view.length()));
```
so it knows to call the `u64` one.

**Comment by LucasChollet** [comment]
It used to be constructed with an initializer-list, so last push was for turning braces into parenthesis. I thought that it would do the trick.

The manual cast will work but I wanted to try my fix first to avoid it. Looks like I have no other choice now, I will add it tomorrow when I go back home.

**Comment by AtkinsSJ** [comment]
Constructing something with `()` or `{}` is the same internally, they just both exist because `()` is ambiguous syntax sometimes.

Not to be confused with a constructor that takes an `initializer_list` argument, or the C-style `{.foo = 1, .bar = 2}` initialization. :cry:

**Comment by LucasChollet** [comment]
My bad, I meant list initialization and not `initializer_list`. And there is a difference, as list initialization doesn't allow narrowing (I had conversion in mind, it's why I tested it).
Anyway, it makes no difference, as converting `size_t` to `u64` is not considered as narrowing.

**Comment by trflynn89** [comment]
This is very much postmortem, but was commit 4ab8ad2ed2817dd7bf92d819f58ed324654b6db7 still valid? `SignedBigInteger` *can't* be negative zero anymore after commit b0d6399f60760e25a55ec9e8e95a1ad322b74b22. Asking because if it became possible again, that's a bug.

**Pull request merged**